spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/TestFunctionCatalog.java

```diff
@@ -0,0 +1,129 @@
+  }
+
+  private static final String[] EMPTY_NAMESPACE = new String[] {};
+  private static final String[] SYSTEM_NAMESPACE = new String[] {"system"};
+  private final FunctionCatalog asFunctionCatalog;
+
+  public TestFunctionCatalog(SparkCatalogConfig catalogConfig) {
+    super(catalogConfig);
+    this.asFunctionCatalog = castToFunctionCatalog(catalogName);
+  }
+
+  @Test
+  public void testListFunctionsViaCatalog() throws NoSuchNamespaceException {
+    Assertions.assertThat(asFunctionCatalog.listFunctions(EMPTY_NAMESPACE))
+        .anyMatch(func -> "iceberg_version".equalsIgnoreCase(func.name()));
+
+    Assertions.assertThat(asFunctionCatalog.listFunctions(SYSTEM_NAMESPACE))
+        .anyMatch(func -> "iceberg_version".equalsIgnoreCase(func.name()));
+  }
+
+  @Test
+  public void testLoadAndDescribeUnboundFunctionsViaCatalog() throws NoSuchFunctionException {
+    for (String[] namespace : ImmutableList.of(EMPTY_NAMESPACE, SYSTEM_NAMESPACE)) {
+      Identifier identifier = Identifier.of(namespace, "iceberg_version");
+      UnboundFunction func = asFunctionCatalog.loadFunction(identifier);
+
+      Assertions.assertThat(func).isExactlyInstanceOf(IcebergVersionFunction.class);
+      Assertions.assertThat(func.name()).asString().isEqualTo("iceberg_version");
+      Assertions.assertThat(func.description())
+          .containsIgnoringCase("Returns the runtime Iceberg version");
+    }
+  }
+
+  @Test
+  public void testBindAndUseFunctionsViaCatalog() throws NoSuchFunctionException {
+    for (String[] namespace : ImmutableList.of(EMPTY_NAMESPACE, SYSTEM_NAMESPACE)) {
+      Identifier identifier = Identifier.of(namespace, "iceberg_version");
+      UnboundFunction unboundFunc = asFunctionCatalog.loadFunction(identifier);
+
+      BoundFunction boundFunc = unboundFunc.bind(new StructType());
+      Assertions.assertThat(boundFunc.getClass().toString())
+          .containsIgnoringCase("IcebergVersionFunctionImpl");
+      Assertions.assertThat(boundFunc.canonicalName()).isEqualTo("iceberg.iceberg_version");
+      Assertions.assertThat(boundFunc.isResultNullable()).isFalse();
+      Assertions.assertThat(boundFunc.isDeterministic()).isTrue();
+
+      // Ensure the bound result produces the expected UTF8String type.
+      ScalarFunction<UTF8String> scalarFunc = (ScalarFunction<UTF8String>) boundFunc;
+      Assert.assertEquals(
+          UTF8String.fromString(IcebergBuild.version()),
+          scalarFunc.produceResult(InternalRow.empty()));
+    }
+  }
+
+  @Test
+  public void testCannotLoadFunctionsFromInvalidNamespace() {
+    String dbNamespace = catalogName + ".db";
+    try {
+      sql("CREATE NAMESPACE IF NOT EXISTS %s", dbNamespace);
+      AssertHelpers.assertThrows(
+          "Function Catalog functions should only be accessible from the system namespace and empty namespace",
+          AnalysisException.class,
+          "Undefined function",
+          () -> sql("SELECT %s.iceberg_version()", dbNamespace));
```

Comment:
I'd also like to have a SQL test that calls `iceberg_version()` because we need to test the end-to-end behavior, not just the catalog implementation directly.

Reply:
That's in `TestIcebergVersionFunction`. There's calls to it in SQL in every possible way.

Reply:
Oh I see your note about moving `TestIcebergVersionFunction` over. Will do.